README.md
NoseGAE: Nose for Google App Engine Testing
==================================================
[![PyPI](https://img.shields.io/pypi/v/NoseGAE.svg)](https://pypi.python.org/pypi/NoseGAE)
[![Downloads](https://img.shields.io/pypi/dm/NoseGAE.svg)](https://pypi.python.org/pypi/NoseGAE)
[![License](https://img.shields.io/pypi/l/NoseGAE.svg)](https://pypi.python.org/pypi/NoseGAE)
[![Build Status](https://travis-ci.org/Trii/NoseGAE.svg?branch=master)](https://travis-ci.org/Trii/NoseGAE)

## Overview

NoseGAE is a [nose](http://nose.readthedocs.org/en/latest/index.html) plugin that makes it easier to
write functional and unit tests for [Google App Engine](https://cloud.google.com/appengine/) applications. 

1. [What does it do?](#what-does-it-do)
	1. [Functional tests](#functional-tests) 
	2. [Unit tests](#unit-tests) 
2. [Configuring the TestBed](#configuring-the-testbed)
	1. [Class Based Tests](#class-based-tests)
	2. [Function Tests](#function-tests)
	3. [Doctest](#doctest)
3. [Changes in 0.5.2](#changes-in-052)
4. [Changes in 0.4.0](#changes-in-040)


## What does it do?

NoseGAE sets up the GAE development environment before your test run. This means that you can easily
write functional tests for your application without having to actually start the dev server and test
over http. The plugin also configures and initializes a TestBed instance, your application_id based off
your `app.yaml`, and sets the proper paths using `dev_appserver.fix_sys_path()`.


### Functional tests

Consider the simple hello world application in `examples/helloworld`:


```python
import webapp2
from jinja2 import Environment


class Hello(webapp2.RequestHandler):
    def get(self):
        self.response.headers['Content-Type'] = 'text/plain'
        env = Environment()
        template = env.from_string("Hello {{ greeting }}!")
        self.response.out.write(template.render(greeting='world'))

app = webapp2.WSGIApplication([('/', Hello)], debug=True)

```

And a simple functional test suite `examples/helloworld/test.py` for the application:

```python
from webtest import TestApp
import unittest
import helloworld

app = TestApp(helloworld.app)


class HelloWorldTest(unittest.TestCase):
    def test_index(self):
        """Tests that the index page for the application

        The page should be served as: Content-Type: text/plain
        The body content should contain the string: Hello world!
        """
        response = app.get('/')
        self.assertEqual(response.content_type, 'text/plain')
        self.assertIn('Hello world!', response.body)

```


Without NoseGAE, the test fails.

```
helloworld$ nosetests --logging-level=ERROR
E
======================================================================
ERROR: Failure: ImportError (No module named webapp2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Josh/Developer/Github/jj/lib/python2.7/site-packages/nose-1.3.4-py2.7.egg/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/Users/Josh/Developer/Github/jj/lib/python2.7/site-packages/nose-1.3.4-py2.7.egg/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/Users/Josh/Developer/Github/jj/lib/python2.7/site-packages/nose-1.3.4-py2.7.egg/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/Users/Josh/Developer/Github/nosegae/examples/helloworld/test.py", line 2, in <module>
    import helloworld
  File "/Users/Josh/Developer/Github/nosegae/examples/helloworld/helloworld.py", line 1, in <module>
    import webapp2
ImportError: No module named webapp2

----------------------------------------------------------------------
Ran 1 test in 0.075s

FAILED (errors=1)
```

With NoseGAE, they pass.


```
helloworld$ nosetests --logging-level=ERROR --with-gae
.
----------------------------------------------------------------------
Ran 1 test in 0.264s

OK

```

### Unit tests

Functional tests are only one kind of test, of course. What if you
want to write unit tests for your data models? Normally, you can't use
your models at all outside of the dev environment, because the Google
App Engine datastore isn't available. However, since the NoseGAE
plugin sets up the development environment around your test run, you
can use models directly in your tests.

Consider the `examples/pets/models.py` file that includes some doctests:

```python
from google.appengine.ext import ndb

class Pet(ndb.Model):
    """The Pet class provides storage for pets.

    >>> # initialize testbed stubs
    >>> testbed.init_memcache_stub()
    >>> testbed.init_datastore_v3_stub()

    You can create a pet:
    >>> muffy = Pet(name=u'muffy', type=u'dog', breed=u"Shi'Tzu")
    >>> muffy # doctest: +ELLIPSIS
    Pet(name=u'muffy', type=u'dog', breed=u"Shi'Tzu", ...)
    >>> muffy_key = muffy.put()

    Once created, you can load a pet by its key:

    >>> muffy_key.get() # doctest: +ELLIPSIS
    Pet(name=u'muffy', type=u'dog', breed=u"Shi'Tzu", ...)

    Or by a query that selects the pet:

    >>> list(Pet.query(Pet.type == 'dog')) # doctest: +ELLIPSIS
    [Pet(name=u'muffy', ...)]

    To modify a pet, change one of its properties and ``put()`` it again.

    >>> muffy_2 = muffy
    >>> muffy_2.age = 10
    >>> muffy_key_2 = muffy_2.put()

    The pet's key doesn't change when it is updated.

    >>> bool(muffy_key == muffy_key_2)
    True
    """
    name = ndb.StringProperty(required=True)
    type = ndb.StringProperty(required=True, choices=("cat", "dog", "bird", "fish", "monkey"))
    breed = ndb.StringProperty()
    age = ndb.IntegerProperty()
    comments = ndb.TextProperty()
    created = ndb.DateTimeProperty(auto_now_add=True, required=True)

    def __repr__(self):
        return ("Pet(name=%r, type=%r, breed=%r, age=%r, "
                "comments=%r, created=%r)" %
                (self.name, self.type, self.breed, self.age,
                 self.comments, self.created))

```

Without NoseGAE, the doctests fail.

```
pets$ nosetests --with-doctest --logging-level=ERROR
F
======================================================================
FAIL: Doctest: models.Pet
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 2201, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for models.Pet
  File "/Users/Josh/Developer/Github/nosegae/examples/pets/models.py", line 4, in Pet

----------------------------------------------------------------------
File "/Users/Josh/Developer/Github/nosegae/examples/pets/models.py", line 15, in models.Pet
Failed example:
    muffy_key = muffy.put()
Exception raised:
    Traceback (most recent call last):
      File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest models.Pet[2]>", line 1, in <module>
        muffy_key = muffy.put()
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/model.py", line 3379, in _put
        return self._put_async(**ctx_options).get_result()
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/tasklets.py", line 325, in get_result
        self.check_success()
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/tasklets.py", line 368, in _help_tasklet_along
        value = gen.throw(exc.__class__, exc, tb)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/context.py", line 810, in put
        key = yield self._put_batcher.add(entity, options)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/tasklets.py", line 371, in _help_tasklet_along
        value = gen.send(val)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/ext/ndb/context.py", line 343, in _put_tasklet
        keys = yield self._conn.async_put(options, datastore_entities)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/datastore/datastore_rpc.py", line 1801, in async_put
        return make_put_call(base_req, pbs, extra_hook)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/datastore/datastore_rpc.py", line 1784, in make_put_call
        service_name=self._api_version)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/datastore/datastore_rpc.py", line 1310, in _make_rpc_call
        rpc = self._create_rpc(config, service_name)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/datastore/datastore_rpc.py", line 1205, in _create_rpc
        rpc = apiproxy_stub_map.UserRPC(service_name, deadline, callback)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/api/apiproxy_stub_map.py", line 414, in __init__
        self.__rpc = CreateRPC(service, stubmap)
      File "/Applications/GoogleAppEngineLauncher.app/Contents/Resources/GoogleAppEngine-default.bundle/Contents/Resources/google_appengine/google/appengine/api/apiproxy_stub_map.py", line 68, in CreateRPC
        assert stub, 'No api proxy found for service "%s"' % service
    AssertionError: No api proxy found for service "datastore_v3"
    
```
     
With NoseGAE, they pass.

```
pets$ nosetests --with-doctest --with-gae
.
----------------------------------------------------------------------
Ran 1 test in 0.228s

OK
```

## Configuring the TestBed

NoseGAE automatically configures your `TestBed` instance for you and then enables any stubs that you
may need for your tests to pass. There are two ways to enable and configure your stubs. The first and
most flexible way is to directly initialize the stub(s) on the `TestBed` instance injected into your
test case. The second, and simpler way, is to set the `nosegae_*` and optional `nosegae_*_kwargs`
attributes on your test case and let NoseGAE configure them for you. The following three sections describe
how to use the `TestBed` in your own tests. 

### Class Based Tests

The simplest of use cases is when your test class extends `unittest.TestCase`. The NoseGAE plugin injects an attribute
named `testbed` to the instance of your test class and configures it based upon any attributes matching the convention
`nosegae_<stubname>` and `nosegae_<stubname>_kwargs`.

This test uses the assigned `testbed` attribute to manually configure each test.

```python
class MyTest(unittest.TestCase):
    def test_using_memcache(self):
        """Unit test using memcache"""
        from google.appengine.api import memcache
        self.testbed.init_memcache_stub()
        memcache.set('test', True, 30)
        self.assertTrue(memcache.get('test'))
    
    def test_using_taskqueue(self):
        """Unit test using the taskqueue"""
        self.testbed.init_taskqueue_stub(root_path='/path/to/app')
        from google.appengine.api import taskqueue
        task_url = '/some/task'
        taskqueue.add(url=task_url)
        taskqueue_stub = self.testbed.get_stub('taskqueue')
        tasks = taskqueue_stub.get_filtered_tasks(url=task_url)
        self.assertEqual(1, len(tasks))
        self.assertEqual(task_url, tasks[0].url)
```

The following test case shows how to write a test that uses the datastore stub based on the simple configuration
method using `nosegae_<stubname>` and `nosegae_<stubname>_kwargs`.

```python
class DataTest(unittest.TestCase):
    # enable the datastore stub
    nosegae_datastore_v3 = True
    
    def test_get_entity(self):
        """Naively tests that we can fetch an entity from the datastore"""
        entity = MyModel.query().get()
        self.assertIsNotNone(entity)
    
```

### Function Tests

This test case uses the `testbed` instance assigned to the function to manually configure any needed stubs.
See `examples/function_manual_config`.

```python
def test_index():
    # test_index.testbed is assigned by the NoseGAE plugin
    test_index.testbed.init_taskqueue_stub(task_retry_seconds=42, root_path=os.path.dirname(__file__))
    # Assume the `/` route fires off a task queue and should pass without exceptions
    app = TestApp(helloworld.app)
    response = app.get('/')
    assert 'Hello world!' in str(response)
```

The following test shows how to use the simple method while passing kwargs to the taskqueue
stub's initialization method. See `examples/issue42_task-queue` for full example code.
 
```python
def test_index():
    # Assume the `/` route fires off a task queue and should pass without exceptions
    app = TestApp(app)
    response = app.get('/')
    assert 'Hello world!' in str(response)

# Enable any stubs needed as attributes off of the test function

# NoseGAE looks for queue.yaml in the root of the
# application when nosegae_taskqueue is True
test_index.nosegae_taskqueue = True

# ...or you can manually set the path and any additional arguments with the kwargs attribute
test_index.nosegae_taskqueue_kwargs = dict(task_retry_seconds=42, root_path=os.path.dirname(__file__))
```

### Doctest

Doctests are a whole other beast. They still work but all `TestBed` configuration has to be done manually.
NoseGAE uses the [nose doctest plugin](http://nose.readthedocs.org/en/latest/plugins/doctests.html) to inject
a global variable named `testbed` into your doctest scope that contains the current active `TestBed` instance.
See `examples/pets/models.py` for full example.

```python
class Pet(ndb.Model):
    """The Pet class provides storage for pets.

    >>> # Initialize stubs using the injected testbed instance
    >>> testbed.init_memcache_stub()
    >>> testbed.init_datastore_v3_stub()

    You can create a pet:
    >>> muffy = Pet(name=u'muffy', type=u'dog', breed=u"Shi'Tzu")
    >>> muffy_key = muffy.put()
    """
    name = ndb.StringProperty(required=True)
    type = ndb.StringProperty(required=True, choices=("cat", "dog", "bird", "fish", "monkey"))
    breed = ndb.StringProperty()
```
## Changes in 0.5.2

The 0.5.2 release introduces preliminary modules support by allowing multiple yaml or paths sent to the
`--gae-application` command line option.

```sh
nosetests --with-gae \
          --gae-application='app.yaml,mobile_frontend.yaml,static_backend.yaml,dispatch.yaml'
```

## Changes in 0.4.0

The 0.4.0 release is a major rewrite to support `dev_appserver2`. This release introduced two important changes listed
below.

### Sandbox is gone

Due to changes in the sandboxing mechanisms in dev_appserver2, it isn't possible for NoseGAE to simulate the deployed
environment any longer. The sandboxing feature had to be removed since there is no longer any way to toggle it between
`nose`s own internal workings.

This means that certain tests may pass locally but the code in question will fail in production due to restricted
modules and functions. As of now there is no workaround but pull requests are welcome!

### Testbed is set up for you

The new plugin automatically sets up the initial `google.appengine.ext.testbed.Testbed` instance and injects it into
your test cases for you.


CHANGELOG
0.5.9
- Put app code ahead of GAE lib in sys.path. Thanks gregpsycle!

0.5.8
- Update the docs, update the tests to use the app engine sdk.

0.5.7
- Re-added code that removes GAE log handler. Thanks obmarg!

0.5.6
- Fixes issue #34 by reloading the google module instead of deleting it from
  sys.modules on setup.

0.5.5
- Fix for taskqueue queue.yaml discovery in the case of --gae-application values that contain yaml files
  example: --gae-application "dispatch.yaml,default/app.yaml,mod1/module.yaml" where queue.yaml is in default/

0.5.4
- Override module stub initialization based on SO post
  http://stackoverflow.com/questions/28166558/invalidmoduleerror-when-using-testbed-to-unit-test-google-app-engine

0.5.3
- Fix regression in automatic taskqueue stup initialization due to app_path becoming a list in 0.5.2
- Add Travis-CI configuration for automated testing so this won't happen again

0.5.2
- Accept multiple modules comma-delimited --gae-application (Issue #22) (see support/modules_example/runtests.sh)
- Prepend GAE lib directory to path instead of append (Issue #23)

0.5.1

- Populate os.environ['CURRENT_VERSION_ID'] using devappserver2's built in
  generation via the 
  google.appengine.tools.devappserver2.application_configuration.ModuleConfiguration
  instance that is part of the ApplicationConfiguration's modules list

0.4.1

- Fix issue with timezone mismatch discovered by viorelsfetea by importing
  google.appengine.tools.os_compat which explicitly sets the timezone to UTC
  so that datetime and time objects created in tests match the ones created
  in the appengine runtime.

0.4.0

- Huge rewrite to support dev_appserver2
- Remove --with-sandbox because it does not work with dev_appserver2
- Automatically configure a TestBed instance and inject it into the test code
  for ease of configuration

0.3.0

- Initial dev_appserver2 support for sdk 1.9.17 via Dmitry Sadovnychyi
  (https://github.com/sadovnychyi/NoseGAE/commit/3452d77cae4e02d536f7eb3907602e3f4ddff16d)

0.2.1

- Updated import for old_dev_appserver for SDK 1.9.1+ (ec4011d2102893569f4f88d7e6c40687cc99754a)
- Added default options patch from gregorynicholas (https://code.google.com/p/nose-gae/issues/detail?id=62)
- Moved code to https://github.com/Trii/NoseGAE since old google code repo is no longer maintained

0.2.0

- Updated hook for compatibility with SDK 1.6.0+ (#59). 

0.1.9

 - Fixed import hook for compatibility with SDK 1.5.1+ (#50)

0.1.8

- A task queue with something other than a default queue can be accessed in your app under test now 
  (bugfix from #42)

0.1.7

- Added support for Google App Engine SDK 1.3.1 (#36).  Actually, the 3rd party lib support is identical 
  to what dev_appserver does now so future lib additions should work seamlessly.
- Worked around setuptools conflicts that may be introduced by globally installed eggs having the 
  google namespace, e.g. protobuf (#34).

0.1.6

- Added option --without-sandbox so that import hooks can be bypassed to 
  use coverage and other edge case modules.  Patch contributed by jgeewax (#26)

0.1.5

- Worked around SDK 1.2.5's way of aggressively adding logging handlers (#25). 
  Note that this still allows Nose to capture and report your app's log messages.
  Thanks to Jason H. Smith for the bug report and patch.

0.1.4

- Added support for Google App Engine SDK 1.1.9 (#18).  Thanks to Mahmoud Arram
  and hackerblinks for patches.
- Fixed a path issue that causes ImportError on Windows (#19).  Thanks to kveretennicov 
  for the patch.
- There is a more informative error now if tests run with something lower than 
  Python 2.5 (#14).

0.1.3

- Fixed bug in handling of modules shared between sandboxed and
  non-sandboxed code that caused sys.modules to be empty in sandbox
  (among other problems). Thanks to Kumar McMillan for the bug report
  (#2).
- Changed sys.path munging to ensure that libs that may be in google
  and system libraries are loaded from google's libdir. Thanks to
  Kumar McMillan for the bug report and patch (#6).
- Fixed bug in how the sandbox shared module set was maintained between
  entries into the sandbox, which could cause a cascade of import errors
  under certain arrangments of test and app code. Thanks to Kumar McMillan
  for the bug report (#7).
  
0.1.2

- Updated datastore handling to use a nosegae-specific datastore by
  default. This prevents the test run from clearing the default
  dev_appserver datastore, as would happen with earlier
  releases. Thanks to Andy for raising the issue.

0.1.1

- Updated self-test to use trestle so that the resulting documentation
  is more clear for end users.


nosegae.py
import os
import logging
import sys
import tempfile
from nose.plugins.base import Plugin
from nose.case import FunctionTestCase

# Solution from
# http://stackoverflow.com/questions/17583443/what-is-the-correct-way-to-share-package-version-with-setup-py-and-the-package
from pkg_resources import get_distribution, DistributionNotFound

try:
    _dist = get_distribution('nosegae')
    # Normalize case for Windows systems
    dist_loc = os.path.normcase(_dist.location)
    here = os.path.normcase(__file__)
    if not here.startswith(os.path.join(dist_loc, 'nosegae')):
        # not installed, but there is another version that *is*
        raise DistributionNotFound
except DistributionNotFound:
    __version__ = 'DEVELOPMENT'
else:
    __version__ = _dist.version

logger = logging.getLogger(__name__)


class NoseGAE(Plugin):
    """Activate this plugin to run tests in Google App Engine dev environment. When the plugin is active,
    Google App Engine dev stubs such as the datastore, memcache, taskqueue, and more can be made available.
    """
    name = 'gae'

    def options(self, parser, env=os.environ):
        super(NoseGAE, self).options(parser, env)
        parser.add_option(
            '--gae-lib-root', default='/usr/local/google_appengine',
            dest='gae_lib_root',
            help='Set the path to the root directory of the Google '
            'Application Engine installation')
        parser.add_option(
            '--gae-application', default=None, action='store', dest='gae_app',
            help='Set the path to the GAE application '
            'under test. Default is the nose `where` '
            'directory (generally the cwd)')
        parser.add_option(
            '--gae-datastore', default=None, action='store', dest='gae_data',
            help='Set the path to the GAE datastore to use in tests. '
            'Note that when using an existing datastore directory, the '
            'datastore will not be cleared before testing begins.')

    def configure(self, options, config):
        super(NoseGAE, self).configure(options, config)
        if not self.enabled:
            return

        if sys.version_info[0:2] < (2, 7):
            raise EnvironmentError(
                "Python version must be 2.7 or greater, like the Google App Engine environment.  "
                "Tests are running with: %s" % sys.version)

        try:
            self._app_path = options.gae_app.split(',')
        except AttributeError:
            self._app_path = [config.workingDir]
        self._data_path = options.gae_data or os.path.join(tempfile.gettempdir(),
                                                           'nosegae.sqlite3')

        if options.gae_lib_root not in sys.path:
            options.gae_lib_root = os.path.realpath(options.gae_lib_root)
            sys.path.insert(0, options.gae_lib_root)

        for path_ in self._app_path:
            path_ = os.path.realpath(path_)
            if not os.path.isdir(path_):
                path_ = os.path.dirname(path_)
            if path_ not in sys.path:
                sys.path.insert(0, path_)

        if 'google' in sys.modules:
            # make sure an egg (e.g. protobuf) is not cached
            # with the wrong path:
            reload(sys.modules['google'])
        try:
            import appengine_config
        except ImportError:
            pass

        # TODO: this may need to happen after parsing your yaml files in
        # The case of modules but I need to investigate further
        import dev_appserver
        dev_appserver.fix_sys_path()  # add paths to libs specified in app.yaml, etc

        # This file handles some OS compat settings. Most notably the `TZ` stuff
        # to resolve https://github.com/Trii/NoseGAE/issues/14.
        # It may need to be removed in the future if Google changes the functionality
        import google.appengine.tools.os_compat

        from google.appengine.tools.devappserver2 import application_configuration

        # get the app id out of your app.yaml and stuff
        self.configuration = application_configuration.ApplicationConfiguration(self._app_path)

        os.environ['APPLICATION_ID'] = self.configuration.app_id
        # simulate same environment as devappserver2
        os.environ['CURRENT_VERSION_ID'] = self.configuration.modules[0].version_id

        self.is_doctests = options.enable_plugin_doctest

        # As of SDK 0.2.5 the dev_appserver.py aggressively adds some logging handlers.
        # This removes the handlers but note that Nose will still capture logging and
        # report it during failures.  See Issue 25 for more info.
        rootLogger = logging.getLogger()
        for handler in rootLogger.handlers:
            if isinstance(handler, logging.StreamHandler):
                rootLogger.removeHandler(handler)

    def startTest(self, test):
        """Initializes Testbed stubs based off of attributes of the executing test

        allow tests to register and configure stubs by setting properties like
        nosegae_<stub_name> and nosegae_<stub_name>_kwargs

        Example

        class MyTest(unittest.TestCase):
            nosegae_datastore_v3 = True
            nosegae_datastore_v3_kwargs = {
              'datastore_file': '/tmp/nosegae.sqlite3,
              'use_sqlite': True
            }

            def test_something(self):
               entity = MyModel(name='NoseGAE')
               entity.put()
               self.assertNotNone(entity.key.id())

        Args
            :param test: The unittest.TestCase being run
            :type test: unittest.TestCase
        """
        from google.appengine.ext import testbed

        self._add_missing_stubs(testbed)

        self.testbed = testbed.Testbed()
        self.testbed.activate()
        # Give the test access to the active testbed
        the_test = test.test
        if isinstance(the_test, FunctionTestCase):
            the_test = the_test.test
        the_test.testbed = self.testbed

        for stub_name, stub_init in testbed.INIT_STUB_METHOD_NAMES.iteritems():
            if not getattr(the_test, 'nosegae_%s' % stub_name, False):
                continue
            stub_kwargs = getattr(the_test, 'nosegae_%s_kwargs' % stub_name, {})
            if stub_name == testbed.TASKQUEUE_SERVICE_NAME:
                self._init_taskqueue_stub(**stub_kwargs)
            elif stub_name == testbed.DATASTORE_SERVICE_NAME:
                if not self.testbed.get_stub(testbed.MEMCACHE_SERVICE_NAME):
                    # ndb requires memcache so enable it as well as the datastore_v3
                    self.testbed.init_memcache_stub()
                self._init_datastore_v3_stub(**stub_kwargs)
            elif stub_name == testbed.USER_SERVICE_NAME:
                self._init_user_stub(**stub_kwargs)
            elif stub_name == testbed.MODULES_SERVICE_NAME:
                self._init_modules_stub(**stub_kwargs)
            else:
                self._init_stub(stub_init, **stub_kwargs)

        if self.is_doctests:
            self._doctest_compat(the_test)
        self.the_test = the_test

    def stopTest(self, test):
        self.testbed.deactivate()
        del self.the_test.testbed
        del self.the_test

    def _doctest_compat(self, the_test):
        """Enable compatibility with doctests by setting the current testbed into the doctest scope"""
        try:
            the_test._dt_test.globs['testbed'] = self.testbed
        except AttributeError:
            # not a nose.plugins.doctests.DocTestCase?
            pass

    def _add_missing_stubs(self, testbed):
        """Monkeypatch the testbed for stubs that do not have an init method yet"""
        if not hasattr(testbed, 'PROSPECTIVE_SEARCH_SERVICE_NAME'):
            from google.appengine.api.prospective_search.prospective_search_stub import ProspectiveSearchStub
            testbed.PROSPECTIVE_SEARCH_SERVICE_NAME = 'matcher'
            testbed.INIT_STUB_METHOD_NAMES.update({
                testbed.PROSPECTIVE_SEARCH_SERVICE_NAME: 'init_prospective_search_stub'
            })

            def init_prospective_search_stub(self, enable=True, data_file=None):
                """Workaround to avoid prospective search complain until there is a proper testbed stub

                http://stackoverflow.com/questions/16026703/testbed-stub-for-google-app-engine-prospective-search

                Args:
                    :param self: The Testbed instance.
                    :param enable: True if the fake service should be enabled, False if real
                        service should be disabled.
                """

                if not enable:
                    self._disable_stub(testbed.PROSPECTIVE_SEARCH_SERVICE_NAME)
                    return
                stub = ProspectiveSearchStub(
                    prospective_search_path=data_file,
                    taskqueue_stub=self.get_stub(testbed.TASKQUEUE_SERVICE_NAME))
                self._register_stub(testbed.PROSPECTIVE_SEARCH_SERVICE_NAME, stub)
            testbed.Testbed.init_prospective_search_stub = init_prospective_search_stub

    def _init_taskqueue_stub(self, **stub_kwargs):
        """Initializes the taskqueue stub using nosegae config magic"""
        task_args = {}
        # root_path is required so the stub can find 'queue.yaml' or 'queue.yml'
        if 'root_path' not in stub_kwargs:
            for p in self._app_path:
                # support --gae-application values that may be a .yaml file
                dir_ = os.path.dirname(p) if os.path.isfile(p) else p
                if os.path.isfile(os.path.join(dir_, 'queue.yaml')) or \
                        os.path.isfile(os.path.join(dir_, 'queue.yml')):
                    task_args['root_path'] = dir_
                    break
        task_args.update(stub_kwargs)
        self.testbed.init_taskqueue_stub(**task_args)

    def _init_datastore_v3_stub(self, **stub_kwargs):
        """Initializes the datastore stub using nosegae config magic"""
        task_args = dict(datastore_file=self._data_path)
        task_args.update(stub_kwargs)
        self.testbed.init_datastore_v3_stub(**task_args)

    def _init_user_stub(self, **stub_kwargs):
        """Initializes the user stub using nosegae config magic"""
        # do a little dance to keep the same kwargs for multiple tests in the same class
        # because the user stub will barf if you pass these items into it
        # stub = user_service_stub.UserServiceStub(**stub_kw_args)
        # TypeError: __init__() got an unexpected keyword argument 'USER_IS_ADMIN'
        task_args = stub_kwargs.copy()
        self.testbed.setup_env(overwrite=True,
                               USER_ID=task_args.pop('USER_ID', 'testuser'),
                               USER_EMAIL=task_args.pop('USER_EMAIL', 'testuser@example.org'),
                               USER_IS_ADMIN=task_args.pop('USER_IS_ADMIN', '1'))
        self.testbed.init_user_stub(**task_args)

    def _init_modules_stub(self, **_):
        """Initializes the modules stub based off of your current yaml files

        Implements solution from
        http://stackoverflow.com/questions/28166558/invalidmoduleerror-when-using-testbed-to-unit-test-google-app-engine
        """
        from google.appengine.api import request_info
        # edit all_versions per modules & versions thereof needing tests
        all_versions = {}  # {'default': [1], 'andsome': [2], 'others': [1]}
        def_versions = {}  # {m: all_versions[m][0] for m in all_versions}
        m2h = {}  # {m: {def_versions[m]: 'localhost:8080'} for m in def_versions}
        for module in self.configuration.modules:
            module_name = module._module_name or 'default'
            module_version = module._version or '1'
            all_versions[module_name] = [module_version]
            def_versions[module_name] = module_version
            m2h[module_name] = {module_version: 'localhost:8080'}

        request_info._local_dispatcher = request_info._LocalFakeDispatcher(
            module_names=list(all_versions),
            module_name_to_versions=all_versions,
            module_name_to_default_versions=def_versions,
            module_name_to_version_to_hostname=m2h)
        self.testbed.init_modules_stub()

    def _init_stub(self, stub_init, **stub_kwargs):
        """Initializes all other stubs for consistency's sake"""
        getattr(self.testbed, stub_init, lambda **kwargs: None)(**stub_kwargs)
